storage/rocksdb/rdb_converter.h

```diff
@@ -0,0 +1,246 @@
+  bool m_decode;
+  // Skip this many bytes before reading (or skipping) this field
+  int m_skip;
+};
+
+/**
+ Class to convert rocksdb value slice from storage format to mysql record
+ format.
+*/
+class Rdb_convert_to_record_value_decoder {
+ public:
+  Rdb_convert_to_record_value_decoder() = delete;
+  Rdb_convert_to_record_value_decoder(
+      const Rdb_convert_to_record_value_decoder &decoder) = delete;
+  Rdb_convert_to_record_value_decoder &
+  operator=(const Rdb_convert_to_record_value_decoder &decoder) = delete;
+
+  static int decode(uchar *const buf, uint *offset, TABLE *table,
+                    my_core::Field *field, Rdb_field_encoder *field_dec,
+                    Rdb_string_reader *reader, bool decode, bool is_null);
+
+ private:
+  static int decode_blob(TABLE *table, Field *field, Rdb_string_reader *reader,
+                         bool decode);
+  static int decode_fixed_length_field(Field *const field,
+                                       Rdb_field_encoder *field_dec,
+                                       Rdb_string_reader *const reader,
+                                       bool decode);
+
+  static int decode_varchar(Field *const field, Rdb_string_reader *const reader,
+                            bool decode);
+};
+
+/**
+  Class to iterator fields in RocksDB value slice
+  A template class instantiation represent a way to decode the data.
+  The reason to use template class instead of normal class is to elimate
+  virtual method call.
+*/
+template <typename value_field_decoder> class Rdb_value_field_iterator {
+ private:
+  bool m_is_null;
+  std::vector<READ_FIELD>::const_iterator m_field_iter;
+  std::vector<READ_FIELD>::const_iterator m_field_end;
+  Rdb_string_reader *m_value_slice_reader;
+  // null value map
+  const char *m_null_bytes;
+  // The current open table
+  TABLE *m_table;
+  // The current field
+  Field *m_field;
+  Rdb_field_encoder *m_field_dec;
+  uchar *const m_buf;
+  uint m_offset;
+
+ public:
+  Rdb_value_field_iterator(TABLE *table, Rdb_string_reader *value_slice_reader,
+                           const Rdb_converter *rdb_converter,
+                           uchar *const buf);
+  Rdb_value_field_iterator(const Rdb_value_field_iterator &field_iterator) =
+      delete;
+  Rdb_value_field_iterator &
+  operator=(const Rdb_value_field_iterator &field_iterator) = delete;
+
+  /*
+    Move and decode next field
+    Run next() before accessing data
+  */
+  int next();
+  // Whether current field is the end of fields
+  bool end_of_fields() const;
+  void *get_dst() const;
+  // Whether the value of current field is null
+  bool is_null() const;
+  // get current field index
+  int get_field_index() const;
+  // get current field type
+  enum_field_types get_field_type() const;
+  // get current field
+  Field *get_field() const;
+};
+
+/**
+  Class to convert Mysql formats to rocksdb storage format, and vice versa.
+*/
+class Rdb_converter {
```

Comment:
Please review these class methods for const correctness.  They may all be OK, but just please place a close eye on it.